include/oneapi/internal/execution_context_sycl.h

```diff
@@ -41,164 +42,77 @@ namespace interface1
-    {}
 
-    void build(ExecutionTargetId target, const char * key, const char * program, const char * options = "",
+    void build(ExecutionTargetId target, const char * name, const char * program, const char * options = "",
                services::Status * status = nullptr) DAAL_C11_OVERRIDE
     {
         // TODO: Thread safe?
-        // TODO Rework of "cache"
-
-        uint64_t id = hash(key) % SIZE_CACHE_PROGRAM;
-
-        while (_clProgramCache[id].getProgram() != nullptr && strcmp(_clProgramCache[id].getName(), key) != 0)
-        {
-            id = (id + 1) % SIZE_CACHE_PROGRAM;
-        }
-
-        if (_clProgramCache[id].getProgram())
+        services::Status localStatus;
+        services::String key = name;
+        const bool res = programHashTable.contain(key, localStatus);
+        DAAL_CHECK_LOCAL_STATUS(status, localStatus)
+        if (!res)
         {
-            _clProgramRef    = _clProgramCache[id].getProgram();
-            _executionTarget = target;
+            auto programPtr = services::SharedPtr<OpenClProgramRef>(new OpenClProgramRef(_deviceQueue.get_context().get(),
+                _deviceQueue.get_device().get(), name, program, options, &localStatus));
+            DAAL_CHECK_LOCAL_STATUS(status, localStatus)
+            programHashTable.add(key, programPtr, localStatus);
+            DAAL_CHECK_LOCAL_STATUS(status, localStatus)
+            _currentProgramRef = programPtr.get();
         }
         else
         {
-            _clProgramCache[id].setProgram(new OpenClProgramRef(_deviceQueue.get_context().get(),
-                                           _deviceQueue.get_device().get(), key, program, options, status));
-            if (status != nullptr && !status->ok())
-            {
-                return;
-            }
-            _clProgramRef    = _clProgramCache[id].getProgram();
-            _executionTarget = target;
+            _currentProgramRef = programHashTable.get(key, localStatus).get();
+            DAAL_CHECK_LOCAL_STATUS(status, localStatus)
         }
+
+        _executionTarget = target;
     }
 
     KernelPtr getKernel(const char * kernelName, services::Status * status = nullptr) DAAL_C11_OVERRIDE
     {
-        DAAL_ASSERT(_clProgramRef);
-        DAAL_ASSERT(*_clProgramRef);
-        KernelPtr kernelPtr;
-
-        services::String keyCache = _clProgramRef->getName();
-        keyCache.add(kernelName);
-        uint64_t id = hash(keyCache.c_str()) % SIZE_CACHE_KERNEL;
-        // TODO: Thread safe?
-
-        while (_kernelCache[id].getKernel() && strcmp(_kernelCache[id].getName(), kernelName) != 0)
+        if (_currentProgramRef == nullptr)
         {
-            id = (id + 1) % SIZE_CACHE_KERNEL;
+            DAAL_CHECK_LOCAL_STATUS_RETURN_IF_FAIL(status, services::Status().add(services::ErrorExecutionContextCache), KernelPtr())
```

Comment:
Can be rewritten this way:
```cpp
services::internal::tryAssignStatus(status, services::ErrorExecutionContextCache);
return KernelPtr();
```

Reply:
Misleading error id  `ErrorExecutionContextCache`. 
Do not introduce `ErrorExecutionContextCache`, `ErrorExecutionContext` would be enough.